protocols/autonatv2/src/server/handler/dial_request.rs

```diff
@@ -0,0 +1,355 @@
+            HandleFail::RequestRejected => Self {
+                status: ResponseStatus::E_REQUEST_REJECTED,
+                addr_idx: 0,
+                dial_status: DialStatus::UNUSED,
+            },
+            HandleFail::DialRefused => Self {
+                status: ResponseStatus::E_DIAL_REFUSED,
+                addr_idx: 0,
+                dial_status: DialStatus::UNUSED,
+            },
+            HandleFail::DialBack { idx, err } => Self {
+                status: ResponseStatus::OK,
+                addr_idx: idx,
+                dial_status: match err {
+                    DialBackStatus::DialErr => DialStatus::E_DIAL_ERROR,
+                    DialBackStatus::DialBackErr => DialStatus::E_DIAL_BACK_ERROR,
+                    DialBackStatus::Ok => DialStatus::OK,
+                },
+            },
+        }
+    }
+}
+
+async fn handle_request_internal<I>(
+    coder: &mut Coder<I>,
+    observed_multiaddr: Multiaddr,
+    dial_back_cmd_sender: mpsc::Sender<DialBackCommand>,
+    mut rng: impl RngCore,
+    all_addrs: &mut Vec<Multiaddr>,
+    tested_addrs: &mut Option<Multiaddr>,
+    data_amount: &mut usize,
+) -> Result<DialResponse, HandleFail>
+where
+    I: AsyncRead + AsyncWrite + Unpin,
+{
+    let DialRequest { mut addrs, nonce } = match coder
+        .next()
+        .await
+        .map_err(|_| HandleFail::InternalError(0))?
+    {
+        Request::Dial(dial_request) => dial_request,
+        Request::Data(_) => {
+            return Err(HandleFail::RequestRejected);
+        }
+    };
+    *all_addrs = addrs.clone();
+    let idx = 0;
+    let addr = addrs.pop().ok_or(HandleFail::DialRefused)?;
+    *tested_addrs = Some(addr.clone());
+    *data_amount = 0;
+    if addr != observed_multiaddr {
+        let dial_data_request = DialDataRequest::from_rng(idx, &mut rng);
+        let mut rem_data = dial_data_request.num_bytes;
+        coder
+            .send(Response::Data(dial_data_request))
+            .await
+            .map_err(|_| HandleFail::InternalError(idx))?;
+        while rem_data > 0 {
+            let data_count = match coder
+                .next()
+                .await
+                .map_err(|_e| HandleFail::InternalError(idx))?
+            {
+                Request::Dial(_) => {
+                    return Err(HandleFail::RequestRejected);
+                }
+                Request::Data(dial_data_response) => dial_data_response.get_data_count(),
+            };
+            rem_data = rem_data.saturating_sub(data_count);
+            *data_amount += data_count;
+        }
+    }
+    let (back_channel, rx) = oneshot::channel();
+    let dial_back_cmd = DialBackCommand {
+        addr,
+        nonce,
+        back_channel,
+    };
+    dial_back_cmd_sender
+        .clone()
+        .send(dial_back_cmd)
+        .await
+        .map_err(|_| HandleFail::DialBack {
+            idx,
+            err: DialBackStatus::DialErr,
+        })?;
+
+    // TODO: add timeout
```

Comment:
Everything within a `FuturesSet` is covered by a timeout so we don't need an additional one here.

Reply:
Good call